README.md
# time for zig. 

This provides functions/api for working with time in ziglang. It is a port of
 go's standard time package, it is not a goal of this library to be `1:1`
 mapping with the go `time` package  API.

## features

- monotonic clock.
- support timezones.
- format time to timestamps.

__WARNING__: This was only tested to work on darwin (mac os), theoretical it might work with linux but don't hold your breath, I don't have a linux machine PR's are welcome though.


This is still in early stage, it isn't feature complete yet. But all important functions are available so you can still have fun and waste your `time` on it.

### Usage

```
const std = @import("std");
const warn = std.debug.warn;
const time = @import("./src/time.zig");

test "now" {
    var now = time.now();

    warn("\n today's date {}", now.date());
    warn("\n today's time {}", now.clock());
    warn("\n local timezone detail  {}\n", now.zone());

    // $ zig test example.zig
    // Test 1/1 now...
    //  today's date DateDetail{ .year = 2018, .month = Month.November, .day = 25, .yday = 328 }
    //  today's time Clock{ .hour = 11, .min = 17, .sec = 16 }
    //  local timezone detail  ZoneDetail{ .name = EAT, .offset = 10800 }
    // OK
    // All tests passed.
}
```

RELEASE_NOTES.md
# Release notes for Time (v0.6.0)

This marks another notable release from `v0.3.1`. For anyone of you who isn't
familiar with what this is all about, [time](https://github.com/gernest/time) is a time package for the zig
programming language. It offers a clean API to manipulate and display time,
this package also supports time zones.

## Notable changes


### Print durations in a human format

`Duration.string` returns a human readable string for the duration.
For instance if you want to print duration of one hour ,four minutes and 10 seconds.

```
    const hour = Duration.Hour.value;
    const minute = Duration.Minute.value;
    const second = Duration.Second.value;
    var d = Duration.init(hour + minute * 4 + second * 10);
    warn("duration is {} \n", d.string());
```
```
duration is 1h4m10s
```


## date manipulations

### Time.addDate

You can add date value to an existing Time instance to get a new date.

For example

```
    var local = time.Location.getLocal();
    var ts = time.now(&local);
    var buf = try std.Buffer.init(std.debug.global_allocator, "");
    defer buf.deinit();
    try ts.string(&buf);
    warn("\ncurrent time is {}\n", buf.toSlice());

    // let's add 1 year
    ts = ts.addDate(1, 0, 0);
    try ts.string(&buf);
    warn("this time next year is {}\n", buf.toSlice());
```

```
current time is 2018-12-08 10:32:30.000178063 +0300 EATm=+419006.837156719
this time next year is 2019-12-08 10:32:30.000178063 +0300 EAT
```

Please play with the api, you can add year,month,day. If you want to go back in
time then use negative numbers. From the example above if you want to back one year then
use `ts.addDate(-1, 0, 0)`


This is just highlight, there are

- lots of bug fixes
- cleanups to more idiomatic zig
- lots of utility time goodies like `Time.beginningOfMonth`
- improved documentation
- lots of tests added
- ... and so much more

I encourage you to take a look at it so you can experiment for yourself.

All kind of feedback is welcome.

Enjoy



src/time_test.zig
// Copyright 2009 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
//
// Copyright 2018 Geofrey Ernest MIT LICENSE

const std = @import("std");
const time = @import("time.zig");
const Location = time.Location;
const mem = std.mem;
const warn = std.debug.warn;

const failed_test = error.Failed;
const January = time.Month.January;
const April = time.Month.April;
const September = time.Month.September;
const December = time.Month.December;
const Monday = time.Weekday.Monday;
const Wednesday = time.Weekday.Wednesday;
const Thursday = time.Weekday.Thursday;
const Saturday = time.Weekday.Saturday;
const Sunday = time.Weekday.Sunday;

const parsedTime = struct {
    year: isize,
    month: time.Month,
    day: isize,
    hour: isize,
    minute: isize,
    second: isize,
    nanosecond: isize,
    weekday: time.Weekday,
    zone_offset: isize,
    zone: []const u8,

    fn init(year: isize, month: time.Month, day: isize, hour: isize, minute: isize, second: isize, nanosecond: isize, weekday: time.Weekday, zone_offset: isize, zone: []const u8) parsedTime {
        return parsedTime{
            .year = year,
            .month = month,
            .day = day,
            .hour = hour,
            .minute = minute,
            .second = second,
            .nanosecond = nanosecond,
            .weekday = weekday,
            .zone_offset = zone_offset,
            .zone = zone,
        };
    }
};

const TimeTest = struct {
    seconds: i64,
    golden: parsedTime,
};

const utc_tests = []TimeTest{
    TimeTest{ .seconds = 0, .golden = parsedTime.init(1970, January, 1, 0, 0, 0, 0, Thursday, 0, "UTC") },
    TimeTest{ .seconds = 1221681866, .golden = parsedTime.init(2008, September, 17, 20, 4, 26, 0, Wednesday, 0, "UTC") },
    TimeTest{ .seconds = -1221681866, .golden = parsedTime.init(1931, April, 16, 3, 55, 34, 0, Thursday, 0, "UTC") },
    TimeTest{ .seconds = -11644473600, .golden = parsedTime.init(1601, January, 1, 0, 0, 0, 0, Monday, 0, "UTC") },
    TimeTest{ .seconds = 599529660, .golden = parsedTime.init(1988, December, 31, 0, 1, 0, 0, Saturday, 0, "UTC") },
    TimeTest{ .seconds = 978220860, .golden = parsedTime.init(2000, December, 31, 0, 1, 0, 0, Sunday, 0, "UTC") },
};

const nano_tests = []TimeTest{
    TimeTest{ .seconds = 0, .golden = parsedTime.init(1970, January, 1, 0, 0, 0, 1e8, Thursday, 0, "UTC") },
    TimeTest{ .seconds = 1221681866, .golden = parsedTime.init(2008, September, 17, 20, 4, 26, 2e8, Wednesday, 0, "UTC") },
};

const local_tests = []TimeTest{
    TimeTest{ .seconds = 0, .golden = parsedTime.init(1969, December, 31, 16, 0, 0, 0, Wednesday, -8 * 60 * 60, "PST") },
    TimeTest{ .seconds = 1221681866, .golden = parsedTime.init(2008, September, 17, 13, 4, 26, 0, Wednesday, -7 * 60 * 60, "PDT") },
};

const nano_local_tests = []TimeTest{
    TimeTest{ .seconds = 0, .golden = parsedTime.init(1969, December, 31, 16, 0, 0, 0, Wednesday, -8 * 60 * 60, "PST") },
    TimeTest{ .seconds = 1221681866, .golden = parsedTime.init(2008, September, 17, 13, 4, 26, 3e8, Wednesday, -7 * 60 * 60, "PDT") },
};

fn same(t: time.Time, u: *parsedTime) bool {
    const date = t.date();
    const clock = t.clock();
    const zone = t.zone();
    const check = date.year != u.year or @enumToInt(date.month) != @enumToInt(u.month) or
        date.day != u.day or clock.hour != u.hour or clock.min != u.minute or clock.sec != u.second or
        !mem.eql(u8, zone.name, u.zone) or zone.offset != u.zone_offset;
    if (check) {
        return false;
    }
    return t.year() == u.year and
        @enumToInt(t.month()) == @enumToInt(u.month) and
        t.day() == u.day and
        t.hour() == u.hour and
        t.minute() == u.minute and
        t.second() == u.second and
        t.nanosecond() == u.nanosecond and
        @enumToInt(t.weekday()) == @enumToInt(u.weekday);
}

test "TestSecondsToUTC" {
    for (utc_tests) |ts| {
        var tm = time.unix(ts.seconds, 0, &Location.utc_local);
        const ns = tm.unix();
        if (ns != ts.seconds) {
            warn("SecondsToUTC({}).Seconds() = {}\n", ts.seconds, ns);
            return failed_test;
        }
        var golden = ts.golden;
        if (!same(tm, &golden)) {
            warn("wrong utc time\n");
            return failed_test;
        }
    }
}

test "TestNanosecondsToUTC" {
    for (nano_tests) |tv| {
        var golden = tv.golden;
        const nsec = tv.seconds * i64(1e9) + @intCast(i64, golden.nanosecond);
        var tm = time.unix(0, nsec, &Location.utc_local);
        const new_nsec = tm.unix() * i64(1e9) + @intCast(i64, tm.nanosecond());
        if (new_nsec != nsec) {
            warn("NanosecondsToUTC({}).Nanoseconds() = {}\n", nsec, new_nsec);
            return failed_test;
        }
        if (!same(tm, &golden)) {
            warn("wrong utc time\n");
            return failed_test;
        }
    }
}

test "TestSecondsToLocalTime" {
    var buf = try std.Buffer.init(std.debug.global_allocator, "");
    defer buf.deinit();
    var loc = try Location.load("US/Pacific");
    defer loc.deinit();
    for (local_tests) |tv| {
        var golden = tv.golden;
        const sec = tv.seconds;
        var tm = time.unix(sec, 0, &loc);
        const new_sec = tm.unix();
        if (new_sec != sec) {
            warn("SecondsToLocalTime({}).Nanoseconds() = {}\n", sec, new_sec);
            return failed_test;
        }
        if (!same(tm, &golden)) {
            try tm.string(&buf);
            warn("wrong local time time {}\n", buf.toSlice());
            return failed_test;
        }
    }
}

test "TestNanosecondsToUTC" {
    var loc = try Location.load("US/Pacific");
    defer loc.deinit();
    for (nano_local_tests) |tv| {
        var golden = tv.golden;
        const nsec = tv.seconds * i64(1e9) + @intCast(i64, golden.nanosecond);
        var tm = time.unix(0, nsec, &loc);
        const new_nsec = tm.unix() * i64(1e9) + @intCast(i64, tm.nanosecond());
        if (new_nsec != nsec) {
            warn("NanosecondsToLocalTime({}).Nanoseconds() = {}\n", nsec, new_nsec);
            return failed_test;
        }
        if (!same(tm, &golden)) {
            warn("wrong NanosecondsToLocalTime\n");
            return failed_test;
        }
    }
}

const formatTest = struct {
    name: []const u8,
    format: []const u8,
    result: []const u8,

    fn init(name: []const u8, format: []const u8, result: []const u8) formatTest {
        return formatTest{ .name = name, .format = format, .result = result };
    }
};

const format_tests = []formatTest{
    formatTest.init("ANSIC", time.ANSIC, "Wed Feb  4 21:00:57 2009"),
    formatTest.init("UnixDate", time.UnixDate, "Wed Feb  4 21:00:57 PST 2009"),
    formatTest.init("RubyDate", time.RubyDate, "Wed Feb 04 21:00:57 -0800 2009"),
    formatTest.init("RFC822", time.RFC822, "04 Feb 09 21:00 PST"),
    formatTest.init("RFC850", time.RFC850, "Wednesday, 04-Feb-09 21:00:57 PST"),
    formatTest.init("RFC1123", time.RFC1123, "Wed, 04 Feb 2009 21:00:57 PST"),
    formatTest.init("RFC1123Z", time.RFC1123Z, "Wed, 04 Feb 2009 21:00:57 -0800"),
    formatTest.init("RFC3339", time.RFC3339, "2009-02-04T21:00:57-08:00"),
    formatTest.init("RFC3339Nano", time.RFC3339Nano, "2009-02-04T21:00:57.0123456-08:00"),
    formatTest.init("Kitchen", time.Kitchen, "9:00PM"),
    formatTest.init("am/pm", "3pm", "9pm"),
    formatTest.init("AM/PM", "3PM", "9PM"),
    formatTest.init("two-digit year", "06 01 02", "09 02 04"),
    // Three-letter months and days must not be followed by lower-case letter.
    formatTest.init("Janet", "Hi Janet, the Month is January", "Hi Janet, the Month is February"),
    // Time stamps, Fractional seconds.
    formatTest.init("Stamp", time.Stamp, "Feb  4 21:00:57"),
    formatTest.init("StampMilli", time.StampMilli, "Feb  4 21:00:57.012"),
    formatTest.init("StampMicro", time.StampMicro, "Feb  4 21:00:57.012345"),
    formatTest.init("StampNano", time.StampNano, "Feb  4 21:00:57.012345600"),
};

test "TestFormat" {
    var tz = try Location.load("US/Pacific");
    defer tz.deinit();
    var ts = time.unix(0, 1233810057012345600, &tz);
    var buf = try std.Buffer.init(std.debug.global_allocator, "");
    defer buf.deinit();
    for (format_tests) |value| {
        try ts.format(&buf, value.format);
        const got = buf.toSlice();
        if (!std.mem.eql(u8, got, value.result)) {
            warn("{}: expected {} got {}\n", value.name, value.result, got);
        }
    }
}

test "Name" {
    time.Time.calendar();
}


src/format.zig
const std = @import("std");
const mem = std.mem;
const warn = std.debug.warn;

// These are predefined layouts for use in Time.Format and time.Parse.
// The reference time used in the layouts is the specific time:
//  Mon Jan 2 15:04:05 MST 2006
// which is Unix time 1136239445. Since MST is GMT-0700,
// the reference time can be thought of as
//  01/02 03:04:05PM '06 -0700
// To define your own format, write down what the reference time would look
// like formatted your way; see the values of constants like ANSIC,
// StampMicro or Kitchen for examples. The model is to demonstrate what the
// reference time looks like so that the Format and Parse methods can apply
// the same transformation to a general time value.
//
// Some valid layouts are invalid time values for time.Parse, due to formats
// such as _ for space padding and Z for zone information.
//
// Within the format string, an underscore _ represents a space that may be
// replaced by a digit if the following number (a day) has two digits; for
// compatibility with fixed-width Unix time formats.
//
// A decimal point followed by one or more zeros represents a fractional
// second, printed to the given number of decimal places. A decimal point
// followed by one or more nines represents a fractional second, printed to
// the given number of decimal places, with trailing zeros removed.
// When parsing (only), the input may contain a fractional second
// field immediately after the seconds field, even if the layout does not
// signify its presence. In that case a decimal point followed by a maximal
// series of digits is parsed as a fractional second.
//
// Numeric time zone offsets format as follows:
//  -0700  ±hhmm
//  -07:00 ±hh:mm
//  -07    ±hh
// Replacing the sign in the format with a Z triggers
// the ISO 8601 behavior of printing Z instead of an
// offset for the UTC zone. Thus:
//  Z0700  Z or ±hhmm
//  Z07:00 Z or ±hh:mm
//  Z07    Z or ±hh
//
// The recognized day of week formats are "Mon" and "Monday".
// The recognized month formats are "Jan" and "January".
//
// Text in the format string that is not recognized as part of the reference
// time is echoed verbatim during Format and expected to appear verbatim
// in the input to Parse.
//
// The executable example for Time.Format demonstrates the working
// of the layout string in detail and is a good reference.
//
// Note that the RFC822, RFC850, and RFC1123 formats should be applied
// only to local times. Applying them to UTC times will use "UTC" as the
// time zone abbreviation, while strictly speaking those RFCs require the
// use of "GMT" in that case.
// In general RFC1123Z should be used instead of RFC1123 for servers
// that insist on that format, and RFC3339 should be preferred for new protocols.
// RFC3339, RFC822, RFC822Z, RFC1123, and RFC1123Z are useful for formatting;
// when used with time.Parse they do not accept all the time formats
// permitted by the RFCs.
// The RFC3339Nano format removes trailing zeros from the seconds field
// and thus may not sort correctly once formatted.
pub const ANSIC = "Mon Jan _2 15:04:05 2006";
pub const UnixDate = "Mon Jan _2 15:04:05 MST 2006";
pub const RubyDate = "Mon Jan 02 15:04:05 -0700 2006";
pub const RFC822 = "02 Jan 06 15:04 MST";
pub const RFC822Z = "02 Jan 06 15:04 -0700"; // RFC822 with numeric zone
pub const RFC850 = "Monday, 02-Jan-06 15:04:05 MST";
pub const RFC1123 = "Mon, 02 Jan 2006 15:04:05 MST";
pub const RFC1123Z = "Mon, 02 Jan 2006 15:04:05 -0700"; // RFC1123 with numeric zone
pub const RFC3339 = "2006-01-02T15:04:05Z07:00";
pub const RFC3339Nano = "2006-01-02T15:04:05.999999999Z07:00";
pub const Kitchen = "3:04PM";
// Handy time stamps.
pub const Stamp = "Jan _2 15:04:05";
pub const StampMilli = "Jan _2 15:04:05.000";
pub const StampMicro = "Jan _2 15:04:05.000000";
pub const StampNano = "Jan _2 15:04:05.000000000";

pub const chunk = enum {
    none,
    stdLongMonth, // "January"
    stdMonth, // "Jan"
    stdNumMonth, // "1"
    stdZeroMonth, // "01"
    stdLongWeekDay, // "Monday"
    stdWeekDay, // "Mon"
    stdDay, // "2"
    stdUnderDay, // "_2"
    stdZeroDay, // "02"
    stdHour, // "15"
    stdHour12, // "3"
    stdZeroHour12, // "03"
    stdMinute, // "4"
    stdZeroMinute, // "04"
    stdSecond, // "5"
    stdZeroSecond, // "05"
    stdLongYear, // "2006"
    stdYear, // "06"
    stdPM, // "PM"
    stdpm, // "pm"
    stdTZ, // "MST"
    stdISO8601TZ, // "Z0700"  // prints Z for UTC
    stdISO8601SecondsTZ, // "Z070000"
    stdISO8601ShortTZ, // "Z07"
    stdISO8601ColonTZ, // "Z07:00" // prints Z for UTC
    stdISO8601ColonSecondsTZ, // "Z07:00:00"
    stdNumTZ, // "-0700"  // always numeric
    stdNumSecondsTz, // "-070000"
    stdNumShortTZ, // "-07"    // always numeric
    stdNumColonTZ, // "-07:00" // always numeric
    stdNumColonSecondsTZ, // "-07:00:00"
    stdFracSecond0, // ".0", ".00", ... , trailing zeros included
    stdFracSecond9, // ".9", ".99", ..., trailing zeros omitted

    stdNeedDate, // need month, day, year
    stdNeedClock, // need hour, minute, second
    stdArgShift, // extra argument in high bits, above low stdArgShift
};

// startsWithLowerCase reports whether the string has a lower-case letter at the beginning.
// Its purpose is to prevent matching strings like "Month" when looking for "Mon".
fn startsWithLowerCase(str: []const u8) bool {
    if (str.len == 0) {
        return false;
    }
    const c = str[0];
    return 'a' <= c and c <= 'z';
}

const chuckResult = struct {
    prefix: []const u8,
    suffix: []const u8,
    chunk: chunk,
};
const std0x = []chunk{
    chunk.stdZeroMonth,
    chunk.stdZeroDay,
    chunk.stdZeroHour12,
    chunk.stdZeroMinute,
    chunk.stdZeroSecond,
    chunk.stdYear,
};

fn nextStdChunk(layout: []const u8) chuckResult {
    var i: usize = 0;
    while (i < layout.len) : (i += 1) {
        switch (layout[i]) {
            'J' => { // January, Jan
                if ((layout.len >= i + 3) and mem.eql(u8, layout[i .. i + 3], "Jan")) {
                    if ((layout.len >= i + 7) and mem.eql(u8, layout[i .. i + 7], "January")) {
                        return chuckResult{
                            .prefix = layout[0..i],
                            .chunk = chunk.stdLongMonth,
                            .suffix = layout[i + 7 ..],
                        };
                    }
                    if (!startsWithLowerCase(layout[i + 3 ..])) {
                        return chuckResult{
                            .prefix = layout[0..i],
                            .chunk = chunk.stdMonth,
                            .suffix = layout[i + 3 ..],
                        };
                    }
                }
            },
            'M' => { // Monday, Mon, MST
                if (layout.len >= 1 + 3) {
                    if (mem.eql(u8, layout[i .. i + 3], "Mon")) {
                        if ((layout.len >= i + 6) and mem.eql(u8, layout[i .. i + 6], "Monday")) {
                            return chuckResult{
                                .prefix = layout[0..i],
                                .chunk = chunk.stdLongWeekDay,
                                .suffix = layout[i + 6 ..],
                            };
                        }
                        if (!startsWithLowerCase(layout[i + 3 ..])) {
                            return chuckResult{
                                .prefix = layout[0..i],
                                .chunk = chunk.stdWeekDay,
                                .suffix = layout[i + 3 ..],
                            };
                        }
                    }
                    if (mem.eql(u8, layout[i .. i + 3], "MST")) {
                        return chuckResult{
                            .prefix = layout[0..i],
                            .chunk = chunk.stdTZ,
                            .suffix = layout[i + 3 ..],
                        };
                    }
                }
            },
            '0' => {
                if (layout.len >= i + 2 and '1' <= layout[i + 1] and layout[i + 1] <= '6') {
                    const x = layout[i + 1] - '1';
                    return chuckResult{
                        .prefix = layout[0..i],
                        .chunk = std0x[x],
                        .suffix = layout[i + 2 ..],
                    };
                }
            },
            '1' => { // 15, 1
                if (layout.len >= i + 2 and layout[i + 1] == '5') {
                    return chuckResult{
                        .prefix = layout[0..i],
                        .chunk = chunk.stdHour,
                        .suffix = layout[i + 2 ..],
                    };
                }
                return chuckResult{
                    .prefix = layout[0..i],
                    .chunk = chunk.stdNumMonth,
                    .suffix = layout[i + 1 ..],
                };
            },
            '2' => { // 2006, 2
                if (layout.len >= i + 4 and mem.eql(u8, layout[i .. i + 4], "2006")) {
                    return chuckResult{
                        .prefix = layout[0..i],
                        .chunk = chunk.stdLongYear,
                        .suffix = layout[i + 4 ..],
                    };
                }
                return chuckResult{
                    .prefix = layout[0..i],
                    .chunk = chunk.stdDay,
                    .suffix = layout[i + 1 ..],
                };
            },
            '_' => { // _2, _2006
                if (layout.len >= i + 4 and layout[i + 1] == '2') {
                    //_2006 is really a literal _, followed by stdLongYear
                    if (layout.len >= i + 5 and mem.eql(u8, layout[i + 1 .. i + 5], "2006")) {
                        return chuckResult{
                            .prefix = layout[0..i],
                            .chunk = chunk.stdLongYear,
                            .suffix = layout[i + 5 ..],
                        };
                    }
                    return chuckResult{
                        .prefix = layout[0..i],
                        .chunk = chunk.stdUnderDay,
                        .suffix = layout[i + 2 ..],
                    };
                }
            },
            '3' => {
                return chuckResult{
                    .prefix = layout[0..i],
                    .chunk = chunk.stdHour12,
                    .suffix = layout[i + 1 ..],
                };
            },
            '4' => {
                return chuckResult{
                    .prefix = layout[0..i],
                    .chunk = chunk.stdSecond,
                    .suffix = layout[i + 1 ..],
                };
            },
            'P' => { // PM
                if (layout.len >= i + 2 and layout[i + 1] == 'M') {
                    return chuckResult{
                        .prefix = layout[0..i],
                        .chunk = chunk.stdPM,
                        .suffix = layout[i + 2 ..],
                    };
                }
            },
            'p' => { // pm
                if (layout.len >= i + 2 and layout[i + 1] == 'm') {
                    return chuckResult{
                        .prefix = layout[0..i],
                        .chunk = chunk.stdpm,
                        .suffix = layout[i + 2 ..],
                    };
                }
            },
            '-' => {
                if (layout.len >= i + 7 and mem.eql(u8, layout[i .. i + 7], "-070000")) {
                    return chuckResult{
                        .prefix = layout[0..i],
                        .chunk = chunk.stdNumSecondsTz,
                        .suffix = layout[i + 7 ..],
                    };
                }
                if (layout.len >= i + 9 and mem.eql(u8, layout[i .. i + 9], "-07:00:00")) {
                    return chuckResult{
                        .prefix = layout[0..i],
                        .chunk = chunk.stdNumColonSecondsTZ,
                        .suffix = layout[i + 9 ..],
                    };
                }
                if (layout.len >= i + 5 and mem.eql(u8, layout[i .. i + 5], "-0700")) {
                    return chuckResult{
                        .prefix = layout[0..i],
                        .chunk = chunk.stdNumTZ,
                        .suffix = layout[i + 5 ..],
                    };
                }
                if (layout.len >= i + 6 and mem.eql(u8, layout[i .. i + 6], "-07:00")) {
                    return chuckResult{
                        .prefix = layout[0..i],
                        .chunk = chunk.stdNumColonTZ,
                        .suffix = layout[i + 6 ..],
                    };
                }
                if (layout.len >= i + 3 and mem.eql(u8, layout[i .. i + 3], "-07")) {
                    return chuckResult{
                        .prefix = layout[0..i],
                        .chunk = chunk.stdNumShortTZ,
                        .suffix = layout[i + 3 ..],
                    };
                }
            },
            'Z' => { // Z070000, Z07:00:00, Z0700, Z07:00,
                if (layout.len >= i + 7 and mem.eql(u8, layout[i .. i + 7], "Z070000")) {
                    return chuckResult{
                        .prefix = layout[0..i],
                        .chunk = chunk.stdISO8601SecondsTZ,
                        .suffix = layout[i + 7 ..],
                    };
                }
                if (layout.len >= i + 9 and mem.eql(u8, layout[i .. i + 9], "Z07:00:00")) {
                    return chuckResult{
                        .prefix = layout[0..i],
                        .chunk = chunk.stdISO8601ColonSecondsTZ,
                        .suffix = layout[i + 9 ..],
                    };
                }
                if (layout.len >= i + 5 and mem.eql(u8, layout[i .. i + 5], "Z0700")) {
                    return chuckResult{
                        .prefix = layout[0..i],
                        .chunk = chunk.stdISO8601TZ,
                        .suffix = layout[i + 5 ..],
                    };
                }
                if (layout.len >= i + 6 and mem.eql(u8, layout[i .. i + 6], "Z07:00")) {
                    return chuckResult{
                        .prefix = layout[0..i],
                        .chunk = chunk.stdISO8601ColonTZ,
                        .suffix = layout[i + 6 ..],
                    };
                }
                if (layout.len >= i + 3 and mem.eql(u8, layout[i .. i + 3], "Z07")) {
                    return chuckResult{
                        .prefix = layout[0..i],
                        .chunk = chunk.stdISO8601ShortTZ,
                        .suffix = layout[i + 6 ..],
                    };
                }
            },
            '.' => { // .000 or .999 - repeated digits for fractional seconds.
                if (i + 1 < layout.len and (layout[i + 1] == '0' or layout[i + 1] == '9')) {
                    const ch = layout[i + 1];
                    var j = i + 1;
                    while (j < layout.len and layout[j] == ch) : (j += 1) {}
                    if (!isDigit(layout, j)) {
                        var st = chunk.stdFracSecond0;
                        if (layout[i + 1] == '9') {
                            st = chunk.stdFracSecond9;
                        }
                        return chuckResult{
                            .prefix = layout[0..i],
                            .chunk = st,
                            .suffix = layout[j..],
                        };
                    }
                }
            },
            else => {},
        }
    }

    return chuckResult{
        .prefix = layout,
        .chunk = chunk.none,
        .suffix = "",
    };
}

fn isDigit(s: []const u8, i: usize) bool {
    if (s.len <= i) {
        return false;
    }
    const c = s[i];
    return '0' <= c and c <= '9';
}

test "chunk" {
    const rs = nextStdChunk("02 yeah");
    warn("{}\n", rs);
}

const long_day_names = [][]const u8{
    "Sunday",
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday",
};

const short_day_names = [][]const u8{
    "Sun",
    "Mon",
    "Tue",
    "Wed",
    "Thu",
    "Fri",
    "Sat",
};

const short_month_names = [][]const u8{
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec",
};

const long_month_names = [][]const u8{
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December",
};

// match reports whether s1 and s2 match ignoring case.
// It is assumed s1 and s2 are the same length.
fn match(s1: []const u8, s2: []const u8) bool {
    if (s1.len != s2.len) {
        return false;
    }
    var i: usize = 0;
    while (i < s1.len) : (i += 1) {
        var c1 = s1[i];
        var c2 = s2[i];
        if (c1 != c2) {
            c1 |= ('a' - 'A');
            c2 |= ('a' - 'A');
            if (c1 != c2 or c1 < 'a' or c1 > 'z') {
                return false;
            }
        }
    }
    return true;
}

fn lookup(tab: [][]const u8, val: []const u8) !usize {
    for (tab) |v, i| {
        if (val.len >= v.len and match(val[0..v.len], v)) {
            return i;
        }
    }
    return error.BadValue;
}


example.zig
const std = @import("std");
const warn = std.debug.warn;
const time = @import("./src/time.zig");
const Duration = time.Duration;

test "now" {
    var local = time.Location.getLocal();
    var now = time.now(&local);

    warn("\n today's date {}", now.date());
    warn("\n today's time {}", now.clock());
    warn("\n local timezone detail  {}\n", now.zone());

    // $ zig test example.zig
    // Test 1/1 now...
    //  today's date DateDetail{ .year = 2018, .month = Month.November, .day = 25, .yday = 328 }
    //  today's time Clock{ .hour = 11, .min = 17, .sec = 16 }
    //  local timezone detail  ZoneDetail{ .name = EAT, .offset = 10800 }
    // OK
    // All tests passed.
}

const formatTest = struct {
    name: []const u8,
    format: []const u8,
    fn init(name: []const u8, format: []const u8) formatTest {
        return formatTest{ .name = name, .format = format };
    }
};

const format_tests = []formatTest{
    formatTest.init("ANSIC", time.ANSIC),
    formatTest.init("UnixDate", time.UnixDate),
    formatTest.init("RubyDate", time.RubyDate),
    formatTest.init("RFC822", time.RFC822),
    formatTest.init("RFC850", time.RFC850),
    formatTest.init("RFC1123", time.RFC1123),
    formatTest.init("RFC1123Z", time.RFC1123Z),
    formatTest.init("RFC3339", time.RFC3339),
    formatTest.init("RFC3339Nano", time.RFC3339Nano),
    formatTest.init("Kitchen", time.Kitchen),
    formatTest.init("am/pm", "3pm"),
    formatTest.init("AM/PM", "3PM"),
    formatTest.init("two-digit year", "06 01 02"),
    // Three-letter months and days must not be followed by lower-case letter.
    formatTest.init("Janet", "Hi Janet, the Month is January"),
    // Time stamps, Fractional seconds.
    formatTest.init("Stamp", time.Stamp),
    formatTest.init("StampMilli", time.StampMilli),
    formatTest.init("StampMicro", time.StampMicro),
    formatTest.init("StampNano", time.StampNano),
};

test "time.format" {
    var local = time.Location.getLocal();
    var ts = time.now(&local);

    var buf = try std.Buffer.init(std.debug.global_allocator, "");
    defer buf.deinit();
    warn("\n");
    for (format_tests) |value| {
        try ts.format(&buf, value.format);
        const got = buf.toSlice();
        warn("{}:  {}\n", value.name, got);
    }

    // Test 2/2 time.format...
    // ANSIC:  Thu Nov 29 05:46:03 2018
    // UnixDate:  Thu Nov 29 05:46:03 EAT 2018
    // RubyDate:  Thu Nov 29 05:46:03 +0300 2018
    // RFC822:  29 Nov 18 05:46 EAT
    // RFC850:  Thursday, 29-Nov-18 05:46:03 EAT
    // RFC1123:  Thu, 29 Nov 2018 05:46:03 EAT
    // RFC1123Z:  Thu, 29 Nov 2018 05:46:03 +0300
    // RFC3339:  2018-11-29T05:46:03+03:00
    // RFC3339Nano:  2018-11-29T05:46:03.000024416+03:00
    // Kitchen:  5:46AM
    // am/pm:  5am
    // AM/PM:  5AM
    // two-digit year:  18 11 29
    // Janet:  Hi Janet, the Month is November
    // Stamp:  Nov 29 05:46:03
    // StampMilli:  Nov 29 05:46:03.000
    // StampMicro:  Nov 29 05:46:03.000024
    // StampNano:  Nov 29 05:46:03.000024416
    // OK
    // All tests passed.
}

test "durations" {
    // print w0ne hour and 10 secods
    const hour = Duration.Hour.value;
    const minute = Duration.Minute.value;
    const second = Duration.Second.value;
    var d = Duration.init(hour + minute * 4 + second * 10);
    warn("duration is {} \n", d.string());
}

test "addDate" {
    var local = time.Location.getLocal();
    var ts = time.now(&local);
    var buf = try std.Buffer.init(std.debug.global_allocator, "");
    defer buf.deinit();
    try ts.string(&buf);
    warn("\ncurrent time is {}\n", buf.toSlice());

    // let's add 1 year
    ts = ts.addDate(1, 0, 0);
    try ts.string(&buf);
    warn("this time next year is {}\n", buf.toSlice());
}
